### PR TITLE
Travis-CI Build - [/retest and /test perticualt JOB] on PR support for CI-BOT

### PR DESCRIPTION
**Pull request by pavan187**
Travis-CI Build Trigger on PR changes for
1. /retest - Retest CI Build on PR changes
2. /test jobname - test CI build on flaky JOB support

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR will the add the feature for Trigger the Build on PR changes and Trigger a perticualr Build.

**Which issue(s) this PR fixes**: #7 

**Special notes for your reviewer**:
None

### CODE REVIEW

**Comment by m1093782566** [comment]
Great! Thanks @pavan187 

cc @edisonxiang

**Comment by edisonxiang** [comment]
@pavan187  Thanks.
I am trying to make every operation(e.g test/reset) in one independent directory. 
Related with pr #8.
Or we will handle lots of operations for the `github.IssueCommentEvent` in one file `issue.go`
https://github.com/Huawei-PaaS/ci-bot/blob/master/handlers/issue.go#L26
like /assign and /unassign, /label and /remove-label, /test and /retest

Could you put these codes into a single directory named [`trigger`](https://github.com/kubernetes/test-infra/tree/master/prow/plugins/trigger) or some other names you like?

**Review by m1093782566 — APPROVED** [review]
LGTM overall but I have few bandwidth in recent days.

Ping @sids-b and @edisonxiang for sharing review burden. Thanks!

File: handlers/ci.go [hunk 1]
```diff
@@ -1,29 +1,179 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		glog.Errorf("fail to read resp: %v", err)
+	}
+	return resp.StatusCode, body
+}
+
+//Function to handle RETEST all jobs on PR changes
+func (s *Server) SendToCIForRetestAllJobs(PRID int) {
+	glog.Info("going to send test request to circle ci")
```
Comment (rohitsardesai83) [inline handlers/ci.go]: circle ci -> travis ci ?

File: handlers/ci.go [hunk 2]
```diff
@@ -1,29 +1,179 @@
...
+	//Fetch the Build id from Requests
+	for _, PrRequest := range TravisRespBody.Requests {
+		for i, _ := range PrRequest.Builds {
+			if PrRequest.Builds[i].PullRequestNumber == PRID {
+				RespBody := GetJobIdsFromTravisBuild(PrRequest.Builds[i].Href, s.Config.TravisCIToken)
+				err = json.Unmarshal(RespBody, &TravisJobRespBody)
+				if err != nil {
+					glog.Errorf("fail to unmarshal: %v", err)
+				}
+				//Range over all jobs to match the JonID and Trigger Build
```
Comment (rohitsardesai83) [inline handlers/ci.go]: JonID -> JobID

File: handlers/ci.go [hunk 3]
```diff
@@ -1,29 +1,179 @@
...
+	Build            = "build"                     //Job name
+	Verify           = "verify"                    //Job name
+	Unittest         = "unittest"                  //Job name
+	Integration      = "integration"               //Job name
+	Crossbuild       = "crossbuild"                //Job name
+
+)
+
+//TravisCI `/requests` API structure response
+type TravisRespStruct struct {
```
Comment (rohitsardesai83) [inline handlers/ci.go]: can we move all such structs to a new file  types.go ?

File: handlers/ci.go [hunk 4]
```diff
@@ -1,29 +1,179 @@
...
 )
 
-type CircleCIInfo struct {
-	CircleJob string `json:"build_parameters[CIRCLE_JOB]"`
-	Revision  string `json:"revision"`
+const (
+	TravisCIEndPoint = "https://api.travis-ci.org" //travis-ci endpoint
+	ContentTypeJSON  = "application/json"          //content-type
+	TravisAPIVersion = "Travis-API-Version"        //API version, a Mandatory header field for Travis-CI V3 APIs
+	Build            = "build"                     //Job name
```
Comment (rohitsardesai83) [inline handlers/ci.go]: these are job types . better we move them to a separate group. something like this : 

```go
type TravisJobType string

const (
    Verify TravisJobType = "verify"
    UnitTest = "unittest"
    .....
)
```

File: handlers/ci.go [hunk 5]
```diff
@@ -1,29 +1,179 @@
...
+func (s *Server) SendToCIForRetestAllJobs(PRID int) {
+	glog.Info("going to send test request to circle ci")
+
+	var Reqbody io.Reader
+	var TravisRespBody TravisRespStruct
+
+	url := fmt.Sprintf("%s/repo/%s/requests", TravisCIEndPoint, s.Config.TravisRepoAccount)
+	req, err := http.NewRequest(http.MethodGet, url, Reqbody)
+	if err != nil {
+		glog.Errorf("HTTP request is failed: %v", err)
```
Comment (rohitsardesai83) [inline handlers/ci.go]: HTTP request is failed -> HTTP request failed.

File: handlers/ci.go [hunk 6]
```diff
@@ -1,29 +1,179 @@
...
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		glog.Errorf("fail to read resp: %v", err)
+	}
+	return resp.StatusCode, body
+}
+
+//Function to handle RETEST all jobs on PR changes
+func (s *Server) SendToCIForRetestAllJobs(PRID int) {
```
Comment (rohitsardesai83) [inline handlers/ci.go]: rename PRID to prID

File: handlers/ci.go [hunk 7]
```diff
@@ -1,29 +1,179 @@
...
+			if PrRequest.Builds[i].PullRequestNumber == PRID {
+				RespBody := GetJobIdsFromTravisBuild(PrRequest.Builds[i].Href, s.Config.TravisCIToken)
+				err = json.Unmarshal(RespBody, &TravisJobRespBody)
+				if err != nil {
+					glog.Errorf("fail to unmarshal: %v", err)
+				}
+				//Range over all jobs to match the JonID and Trigger Build
+				for _, job := range TravisJobRespBody.Jobs {
+					JobId := strings.Split(job.Number, ".")[1]
+					if Build == jobname && JobId == "1" {
```
Comment (rohitsardesai83) [inline handlers/ci.go]: are these JobId's constants ? can we define them as iota constants ?

File: handlers/issue.go [hunk 8]
```diff
@@ -50,6 +51,24 @@ func (s *Server) handleIssueCommentEvent(body []byte, client *github.Client) {
 	//operation is the assign or unassign check
 	operation := strings.Trim(substrings[0], " ")
 
+	// https://github.com/islinwb/test/pull/1
+	prID := strings.SplitAfter(commentEvent.Issue.PullRequestLinks.GetHTMLURL(), "github.com/")[1]
+
+	//First index contains -islinwb/test/pull/1
+	SubStrings := strings.Split(prID, "/")
```
Comment (rohitsardesai83) [inline handlers/issue.go]: SubStrings    -> subStrings.

File: handlers/issue.go [hunk 9]
```diff
@@ -50,6 +51,24 @@ func (s *Server) handleIssueCommentEvent(body []byte, client *github.Client) {
...
+	len := len(SubStrings)
+	PRNum, err := strconv.Atoi(SubStrings[len-1])
```
Comment (rohitsardesai83) [inline handlers/issue.go]: PRNum -> prNum

File: handlers/ci.go [hunk 10]
```diff
@@ -1,29 +1,179 @@
...
+		glog.Errorf("fail to read resp: %v", err)
+	}
+	return resp.StatusCode, body
+}
+
+//Function to handle RETEST all jobs on PR changes
+func (s *Server) SendToCIForRetestAllJobs(PRID int) {
+	glog.Info("going to send test request to circle ci")
+
+	var Reqbody io.Reader
```
Comment (rohitsardesai83) [inline handlers/ci.go]: all local variables better avoid with capitalized letter at the beginning. applies to all the code throughout.

**Review by rohitsardesai83 — COMMENTED** [review]
please see review comments

**Comment by rohitsardesai83** [comment]
please resolve conflicts. thanks.

**Comment by pavan187** [comment]
> @pavan187 Thanks.
> I am trying to make every operation(e.g test/reset) in one independent directory.
> Related with pr #8.
> Or we will handle lots of operations for the `github.IssueCommentEvent` in one file `issue.go`
> https://github.com/Huawei-PaaS/ci-bot/blob/master/handlers/issue.go#L26
> like /assign and /unassign, /label and /remove-label, /test and /retest
> 
> Could you put these codes into a single directory named [`trigger`](https://github.com/kubernetes/test-infra/tree/master/prow/plugins/trigger) or some other names you like?

Yes, Will be doing it

**Pull request closed without merging**